extend torch.jit._overload to module methods

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24259 extend torch.jit._overload to module methods**

Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Differential Revision: [D16921304](https://our.internmc.facebook.com/intern/diff/D16921304)

Commit:
extend torch.jit._overload to modules

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

driazati commented on test/test_jit.py:
```diff
@@ -13201,14 +13196,61 @@ def __init__(self):
+                return self.hello("hi"), self.hello(.5)
+
+        w = CompileOverloadError()
+        with self.assertRaisesRegex(Exception, "but instead found type \'str\'"):
```
Shouldn't a `str` work here since there's an overload for it?

eellison replied:
This is an error here because you're trying to do `x + 1` with a str

driazati commented on torch/_jit_internal.py:
```diff
@@ -276,6 +278,50 @@ def _get_fn_overloads(qual_name):
+        class_name_map = {}
+        _overloaded_methods[qual_name] = class_name_map
+
+    class_name = get_class_name(func)
```
This doesn't seem like a safe way to globally identify classes since it will break for nested classes with the same class name and same function names since the name is not fully qualified

eellison replied:
Sure, i can add a check that the line number is the same. This is still internal-only tho.

driazati commented on torch/jit/_recursive.py:
```diff
@@ -117,19 +117,48 @@ def recursive_script(mod):
+
     methods = methods + tuple(exported)
 
+    overload_name_mappings = dict(getattr(mod, "__overloads__", {}))
```
Can you delete the `__overloads__` and replace it with these new decorators instead of adding both together? (or do that in a stacked PR)

eellison replied:
yea i was going to do that as a follow up, since that likely breaks the ScriptModules in quantization

driazati commented on torch/jit/_recursive.py:
```diff
@@ -117,19 +117,48 @@ def recursive_script(mod):
+
+    for orig_fn, overload_fns in overloads:
+        orig_ast = torch.jit.get_jit_def(orig_fn, self_name="ScriptModule")
+        names = list(map(lambda i: orig_ast.name().name + "__" + str(i), range(len(overload_fns))))
```
Since the overloaded methods aren't directly callable why do they need to be added as normal methods with this mangled name to the `ScriptModule`? The class could hold its schemas for each overload and check that in `OverloadedMethodValue` instead of what it's doing now

eellison replied:
The normal name isn't being added, only the mangled ones are being added. 

Defining the mangled ones lazily would be an improvement but I don't think it's necessary for this PR. It would also complicate export logic / possibly other things, and would require removing the other `__overload__` mechanism first.

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Differential Revision: [D16921304](https://our.internmc.facebook.com/intern/diff/D16921304)

Commit:
Update on "extend torch.jit._overload to module methods"


Follow up to https://github.com/pytorch/pytorch/pull/23886, add the same overload api specified in PEP 484 to module methods to reduce the friction of adding method overloads that was brought up in #23266.

The usage is:
```
@torch.jit.overload
def add(self, y: int) -> int: ...
@torch.jit.overload
def add(self, y: float) -> float: ...
def add():
   ... 
```

Differential Revision: [D16921304](https://our.internmc.facebook.com/intern/diff/D16921304)
